src/rewrites.rs

```diff
@@ -849,27 +1037,519 @@ impl MultiPatterns {
+                                let (nodes_in_1, _) = add_newly_added(
+                                    rule.2.ast.as_ref(),
+                                    &mut runner.egraph,
+                                    &merged_subst,
+                                    &existing_1,
+                                );
+                                let existing_2_updated: HashSet<Mdl> = existing_2
+                                    .iter()
+                                    .chain(nodes_in_1.iter())
+                                    .map(|node| node.clone())
+                                    .collect();
+                                add_newly_added(
+                                    rule.3.ast.as_ref(),
+                                    &mut runner.egraph,
+                                    &merged_subst,
+                                    &existing_2_updated,
+                                );
+                            }
+
+                            runner.egraph.union(id_1, match_1.eclass);
                             runner.egraph.union(id_2, match_2.eclass);
+                        } else {
                         }
                     }
                 }
             }
         }
     }
+
+    /// Returns true if there will not be a cycle introduced by applying this rule.
+    ///
+    /// Checking based on descendents collected at beginning of run_one(). If any input node
+    /// contains descendents that is any of the matched output class, then there can be a cycle
+    /// created.
+    ///
+    /// # Parameters
+    ///
+    /// - `egraph`: egraph of interest
+    /// - `input_subst`: substitution containing the input variables
+    /// - `var_map_1`: keys of this map contains all the input variables in source pattern 1
+    /// - `var_map_2`: keys of this map contains all the input variables in source pattern 2
+    /// - `out_class_1`: Id of the matched eclass of the output of source pattern 1
+    /// - `out_class_2`: Id of the matched eclass of the output of source pattern 2
+    fn check_cycle_partial(
+        &self,
+        egraph: &EGraph<Mdl, TensorAnalysis>,
+        input_subst: &Subst,
+        var_map_1: &HashMap<egg::Var, egg::Var>,
+        var_map_2: &HashMap<egg::Var, egg::Var>,
+        out_class_1: Id,
+        out_class_2: Id,
+    ) -> bool {
+        // Get all input eclass IDs
+        let input_ids: HashSet<Id> = var_map_1
+            .iter()
+            .chain(var_map_2.iter())
+            .map(|(var, _)| *input_subst.get(*var).unwrap())
+            .collect();
+        // Check descendents of the input eclasses
+        for id in input_ids.iter() {
+            let descendents = self.descendents.as_ref().unwrap();
+            let descendents_input = descendents.get(id).unwrap();
+            if descendents_input.contains(&out_class_1) || descendents_input.contains(&out_class_2)
+            {
+                return false;
+            }
+        }
```

Comment:
```
return input_ids.iter().all(|id| {
  let descendents = self.descendents.as_ref().unwrap();
  let descendents_input = descendents.get(id).unwrap();
  !descendents_input.contains(&out_class_1) && !descendents_input.contains(&out_class_2)
});
```

It also looks to me like you could move the first condition out of the loop.

Reply:
Done